lib/ansible/modules/pip.py

```diff
@@ -611,6 +612,46 @@ def setup_virtualenv(module, env, chdir, out, err):
     return out, err, cmd
 
 
+def _resolve_package_names(
+        module: AnsibleModule,
+        package_list: list[Package],
+        pip: list[str],
+        python_bin: str,
+) -> list[Package]:
+    """Return a list of Packages with names extracted from metadata using pip"""
+    pkgs_to_resolve = [pkg for pkg in package_list if not pkg.has_requirement]
+
+    if not pkgs_to_resolve:
+        return package_list
+
+    # pip install --dry-run is not available in pip versions older than 22.2 and it doesn't
+    # work correctly on all cases until 24.1, so check for this and use the non-resolved
+    # package names if pip is outdated.
+    pip_dep = _get_package_info(module, "pip", python_bin)
+
+    installed_pip = LooseVersion(pip_dep.split('==')[1])
+    minimum_pip = LooseVersion("24.1")
+
+    if installed_pip < minimum_pip:
+        module.warn("Using check mode with packages from vcs urls, file paths, or archives will not behave as expected when using pip versions <24.1.")
+        return package_list  # Just use the default behavior
+
+    with tempfile.NamedTemporaryFile() as tmp_file:
+        module.run_command(
+            [*pip, 'install', '--dry-run', '--ignore-installed', f'--report={tmp_file.name}', *(str(pkg) for pkg in pkgs_to_resolve)],
+            check_rc=True,
+        )
+        report = json.load(tmp_file)
```

Comment:
I somehow missed that for some reason we write to disk unnecesarily. Let's keep everything in-memory:
```suggestion
    _rc, b_meta_json, _stderr = module.run_command(
        (
            *pip, 'install',
            '--dry-run',
            '--ignore-installed',
            '--report=-',
            *map(str, pkgs_to_resolve),
        ),
        check_rc=True,
    )
    report = json.loads(to_native(b_meta_json))
```

Reply:
That's good with me (also fine I'll acquiesce on the map(str) lol)

Reply:
Just a quick note on this - The tmpfile was in use because when it was added there were environment variables for things like `NO_COLOR` and we had the `--quiet` flag. If the issue of parsing output with color sequences is no longer an issue then the current state (no tmpfile to stay in-memory + the inclusion of `--quiet`) should be good.

Reply:
@pkingstonxyz ah, I missed when this PR lost the env vars — they should still be set correctly.

Reply:
I think we should keep the temp file, since that will completely avoid the possibility of unexpected output being mixed in with the report. I'm concerned that there could be warning messages or other output which could corrupt the data we're expecting. The overhead of the tempfile compared to all the other work going on is trivial.

Reply:
The warnings go into stderr and not stdout so it's pretty much impossible. But okay, we could keep the tmp file provided that we do have a test case for env vars messing with ANSI escapes.